### PR TITLE
iOS changes

### PR DESCRIPTION
**Pull request by jpalao**
Rebased to v.5.32.0, substitutes #17879

### CODE REVIEW

**Comment by Leont** [comment]
This PR is extremely big at 46kLOC's, mainly because of canned configurations.

Is that really necessary? This is quite different from other cross-compiled targets (like Android)

**Comment by jpalao** [comment]
> This PR is extremely big at 46kLOC's, mainly because of canned configurations.
>
> Is that really necessary? This is quite different from other cross-compiled targets (like Android)
>
Thank you Leon, yes this is acknowledged. All of the changing bits could be tokenized and the file reduced to one
About Policy.sh at the moment I am patching that with the prefix only

> I don't think this file should have been committed?

this may not be necessary and in the same process we may get rid of it

Following is the introductory message that I failed to send yesterday. I was fixing some manifest and pod problems, and I was not expecting such quick reaction :)

PR below has been rebased to blead and recreated, was https://github.com/Perl/perl5/pull/17879

===

Dear perl5 porters,

Lately I have been investigating how to build perl5 for the iOS family of Apple OSs (darwinvariants other than i386, x86_64 and PowerPC, covered by the existing darwin port)


I thought I would share this experimental patchearly in the hope that this interests you. This was forked from maint-5.30 branch, checked out v5.30.2, wrote the patches and then rebased to v5.30.3.

Pull request is here:

https://github.com/Perl/perl5/pull/17880


ios/build.sh will cross compile perl5 targeting Apple's iOS, tvOS and watchOS and their x86_64 emulators. I have not tested yet the older i386 emulators but these should work with little effort


It comes with some pre-cooked config files, and some patches to darwin hints. A lot of the configurations are common and these files could be thinned

Targets architectures include x86_64, armv7, arm64 and armv7k
The build process completes through 'make test_prep', skips 'make test' and does 'make install'. For testing I am working on an app to run the harness in the device or simulator


In the case of iOS,darwin.sh is the only modified source file of the perl5 tree. The build scriptpatches all other files on the fly


Apple Watch or Apple TV builds require a few additional changes,  I will be pushing these later separately after cleaning up. iOS builds should work already


Any feedback about the approach taken or any suggestions are very much appreciated

Thank you,


Jose Palao

**Comment by Leont** [comment]
> For testing I am working on an app to run the harness in the device or simulator

That is pretty much how cross-compiling works on other architectures, in fact manage to actually run Configure that way.

Such an approach would likely result in less maintenance in the future.

**Comment by jpalao** [comment]
yes I understand this Leon, however there is no shell in the device to run Configure, also no exec, no fork no backticks operator etc
but we can build perl, embed it the test app, build and run the test suite, then parse the results

**Comment by jpalao** [comment]
all targets build now with the latest set of changes

**Comment by Leont** [comment]
Possibly, [perl-cross](https://github.com/arsv/perl-cross/) can help here?

**Comment by khwilliamson** [comment]
@jpalao this needs to be rebased

**Comment by jpalao** [comment]
I have rebased the PR to current blead, Thx

**Comment by jpalao** [comment]
I think I have a nice solution now for running t/harness on both the device and simulator. Looks promising though still needs some work. A few tests are failing, etc

For this purpose I have been using an app that is not suitable for this PR as it would make it even bigger and has a lot of unrelated functionality, so I will be refactoring it into a minimal app that does run the script and outputs to screen and to a file that can then be tailed from the building host for simulator builds or retrieved from the device after tests finish. 

Also I have written an additional module and XS code to handle tests subprocess spawning and capture to use as runperl. This currently depends on my CamelBones iOS fork, so this module is required at the moment on iOS builds to run the tests.

For fork emulation I also have written patches that I will integrate and test once the current test suite is passing

Last step for a nice port would then be to minimize the canned configurations

However I have significantly less time now than I had when I started the PR so please be patient :)

Your feedback is most welcome
Thx

**Comment by jkeenan** [comment]
> I think I have a nice solution now for running t/harness on both the device and simulator. Looks promising though still needs some work. A few tests are failing, etc
> 
> For this purpose I have been using an app that is not suitable for this PR as it would make it even bigger and has a lot of unrelated functionality, so I will be refactoring it into a minimal app that does run the script and outputs to screen and to a file that can then be tailed from the building host for simulator builds or retrieved from the device after tests finish.
> 
> Also I have written an additional module and XS code to handle tests subprocess spawning and capture to use as runperl. This currently depends on my CamelBones iOS fork, so this module is required at the moment on iOS builds to run the tests.
> 
> For fork emulation I also have written patches that I will integrate and test once the current test suite is passing
> 
> Last step for a nice port would then be to minimize the canned configurations
> 
> However I have significantly less time now than I had when I started the PR so please be patient :)
> 
> Your feedback is most welcome
> Thx

This p.r. has languished since last July.  To move it forward, we would need feedback from two types of contributors (perhaps overlapping): (a) people familiar with iOS; (b) people familiar with longer-term maintenance of cross-compiled code.  Any volunteers?

Thank you very much.
Jim Keenan

**Comment by jpalao** [comment]
> This p.r. has languished since last July. To move it forward, we would need feedback from two types of contributors (perhaps overlapping): (a) people familiar with iOS; (b) people familiar with longer-term maintenance of cross-compiled code. Any volunteers?

I apologize for the late reply Jim. Rebasing to latest blead seems to still build. I can push the changes to this or another (better named) branch. Also to some extent I think I will be able to continue working on the harness app and other stuff in the next days

While I am no expert in any of these areas, regarding (a) I have a couple of paid support tickets that I am willing to offer for this purpose and regarding (b)  please let me know if/how I can help with this, I have some scripts that might help

Thx :)

**Comment by Leont** [comment]
> (b) people familiar with longer-term maintenance of cross-compiled code. 

These canned config files are not maintainable. What probably would be maintainable is having the infrastructure to generate them.

**Comment by jpalao** [comment]
> > (b) people familiar with longer-term maintenance of cross-compiled code.
> 
> These canned config files are not maintainable. What probably would be maintainable is having the infrastructure to generate them.

HI Leon, I have reduced the size of the patch to around 20% and pushed the changes to the same branch. I hope this at least alleviates the size problem. Stripping out the c comments from the config.h template (last commit) saves 140K but the resulting file is not as pretty. Thx.

File: embed.h [hunk 1]
```diff
@@ -1498,7 +1498,11 @@
 #    endif
 #  endif
 #  if !(defined(PERL_DEFAULT_DO_EXEC3_IMPLEMENTATION))
+#    if TARGET_OS_IPHONE
```
Comment (Leont) [inline embed.h]: Why is this necessary? I can't imagine iPhone not having exec.

Comment (Leont) [reply embed.h]: And even if it is necessary, it should probably be wired the same way Windows/DOS/OS2's `Perl_do_exec`s are wired.

Comment (jpalao) [reply embed.h]: Thank you for spotting this Leon, I have pushed changes to leave embed.h intact

**Comment by Leont** [comment]
It would help a lot for reviewability if this branch was refactored such that it doesn't contain changes that are reverted in a later branch.

**Comment by Leont** [comment]
Also, it still had hardcoded paths in a bunch of places that contain a version number (in particular 5.30). The version should be substituted in such places.

**Comment by jpalao** [comment]
```
Hello Leon,

I have added harness.app to run t/harness on the device and a
script that I am currently using for testing: ios/test.sh. The idea is to
provide `make test` functionality for ios/build.sh. The documentation now
includes test instructions and a brief overview of the proposed test process.

I have separated the changes which are specific to making the test suite run
and are meant for discussion (mainly in t/) in a new branch here (no PR yet) 

https://github.com/jpalao/perl5/tree/ios_blead_test

So I am testing the branch ios_blead_test which rebases onto ios_blead/ios_5.32.0
Please see hardcoded repo addresses in ios/test.sh

Test Problems
=============

I think most of the problems I have seen stem from the limitations that iOS
has with regards to plain darwin: no tty, no stdin, no shell (and thus
no qx, backticks, system, exec, popen, etc.), one process allowed per
app (no perl runs on main thread 1), binary signature checks,
need for re-lininking, etc. However there may be also problems in how I
implemented things for testing, your feedback is very much appreciated

Also please note that stats collection is broken due to the butcher style patch
on Iterator::Process.pm, your feedback on the testing approach is very much
appreciated.

Example 1
=========

the frequent construct:

    @INC = ...

will not work on iOS when trying to dlopen a .bundle object in the Perl
build dir. These objects are removed after perl build dir is copied to
harness.app Documents dir because harness.app can only load signed
binaries within its own installation path. 

This problem is resolved by using instead:

    use lib ...

Example 2
=========

perl -e is "emulated", so tests will fail with an error 

#      got "Can\'t emulate -E on #! line at /private/var/mobile/Containers/Data/Application/E317E6E6-7B34-4A15-A08D-6C7E2228F566/Documents/t/2EE5B100-7538-42A0-9CAB-D8C4642F007B-389-00000053A7C18B99 line 1.\n"
# expected "Can\'t emulate -E on #! line at -e line 1.\n"

Example 3
=========

Another example is where the eval number differ, like this one in uni/parser.t:

not ok 57 - no panic in pad_findmy_pvn (\#134061)
ok 58 - \#line directive and qq with uni delims inside heredoc
# Failed test 57 - no panic in pad_findmy_pvn (\#134061) at ./test.pl line 1107
#      got "syntax error at (eval 45) line 1, at EOF"
# expected "syntax error at (eval 1) line 1, at EOF"
# PROG: 
# use utf8;
# eval "sort \x{100}%";
# die $@;
# STATUS: 0

===============================================

I hope this helps moving forward.

There are still some instabilities so the test complete test suite usually does not run complete.
Here is a sample complete run

https://www.sendspace.com/file/inf6lu

Jose Palao

```

**Comment by jpalao** [comment]
> Also, it still had hardcoded paths in a bunch of places that contain a version number (in particular 5.30). The version should be substituted in such places.

Thx Leon, I have pushed additional changes to remove the hardcoded values in the templates.

**Comment by jpalao** [comment]
Hello Leon,

I have fixed the problems mentioned in my previous post and tests are passing on 32bit (1.117 M) and 64 bit (1.122 M) arm iOS builds:

* Debugged the testing code in camelbones and cbrunperl.pm. 

* Added osname as darwin-ios, darwin-ios-tv, darwin-ios-watch

* Created t/ios_harness wrapper of t/harness
    @ios_blacklist contains the list of test files currently not passing
    @simul_blacklist contains a list of banned tests that do not run on the simulator (gets added to @ios_blacklist)

* A few tests are already merged onto ios_blead as they contain no other changes than those for the =~ 'darwin' regex change. The remaining test changes are on ios_test_blead. 

* Skipped and documented tests in order to get tests passing.
    For those tests where I think I understand the cause of the failure, I have added an explicit comment.
    In cases have not investigated the cause of failure, or I do not understand the cause, or I do not know how to propose a proper fix, I have tried to note the skips with /iOS.*TODO/  

* Rebased the PRs ios_blead and ios_blead_test to 5.35.4 (around a week ago)
 
 
Currently I am trying to reduce the ios_blead_test branch and checking that no earlier unnecessary changes remain


These are some pending items that could in my opininon reduce the number of changes and/or make for a better port:

* running the tests in the Apple TV and Watch still remains theoretical. The Camelbones framework and harness.app need to be ported still

* pseudofork implementation, this could avoid patching quite some tests

* overloading the backtick operator so that another perl (only) subprocess can be spawned: could avoid patching some tests but since there is no shell it feels like less organic change than the previous, maybe this is not possible?

* integrating 'make' could enable CPAN module installs. I have built GNU 'make' as a dynamic library for iOS, not yet integrated or tested. There is also a PP module that I could test

* extracting testing code camelbones fork and move it to the harness.app

* rename harness.app to perl.app, this could avoid patching some tests

* integrate build with github

* iOS has some CPU limit heuristics built in and harness.app is sometimetimes 'randomly' killed by the OS ...


Also Leon if you and/or maybe other team members are available for a quick chat, I think it would be definitely help


Thx,
Jose

**Comment by jkeenan** [comment]
There has been no progress in this p.r. in 9 months, and the p.r. has acquired merge conflicts.  I recommend that this p.r. be closed and a fresh one be opened when appropriate.

**Comment by jkeenan** [comment]
> There has been no progress in this p.r. in 9 months, and the p.r. has acquired merge conflicts. I recommend that this p.r. be closed and a fresh one be opened when appropriate.

There has been no further discussion of this p.r. in the past 2-1/2+ months, so I'm closing this ticket now.  Please open a new p.r. when needed, including therein a link to this ticket.

**Pull request closed without merging**